### PR TITLE
Allow MIR pass `UnreachableProp` when coverage is enabled

### LINKED ISSUE #116171: Make coverage robust against MIR optimizations removing all counter-increment statements
When coverage instrumentation and MIR opts are both enabled, coverage relies on two assumptions:

- MIR opts that would delete `StatementKind::Coverage` statements instead move them into bb0 and change them to `CoverageKind::Unreachable`.

- MIR opts won't delete all `CoverageKind::Counter statements` from an instrumented function.

The first assumption will be lifted by #116046, which stores coverage mappings in a separate per-function struct, so the original mappings will be available during codegen even if MIR opts delete arbitrary coverage statements.

---

The second assumption is trickier. If a function is instrumented for coverage, but MIR opts remove *all* of its counter-increment statements (e.g. because bb0's ends with `TerminatorKind::Unreachable`), then we will codegen a function with no `llvm.instrprof.increment` intrinsics. This causes LLVM to assume that the function wasn't instrumented, so it will disappear from the final coverage mappings and won't participate in coverage reports.

Coverage has special code for handling functions that are deemed “unused” at the MIR level, but if a function *is* used (e.g. as a function pointer) but its body is unreachable, then that special code isn't applied.

---

Ideally what we need is some way for coverage codegen to detect that an instrumented function has no remaining counter-increment statements, and take special action to either re-insert at least one counter-increment (so LLVM treats it as instrumented), or treat it as though it were unused (so that it shows up in reports as never-executed).

---

@rustbot label +A-code-coverage +T-compiler

### PR DESCRIPTION
**Pull request by Zalathar**
Follow-up to this pass being disabled for coverage in #116166; fixes #116171.

This optimization pass can potentially remove all counter-increment statements from an instrumented function. Previously that would cause it to disappear from coverage reports, so the pass had to be turned off during coverage builds.

This PR introduces a more robust solution: we add an additional post-optimization MIR pass that detects when an instrumented function has lost all of its counter-increment statements, and simply adds another counter-increment statement. That allows coverage codegen to emit the function's coverage mappings as normal.

### CODE REVIEW

**Comment by Zalathar** [comment]
cc @cjgillot for awareness because you have been working on this pass, though I don't think this change directly affects your work

File: compiler/rustc_mir_transform/src/coverage/repair.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use rustc_span::DUMMY_SP;
+
+/// If a function has been [instrumented for coverage](super::InstrumentCoverage),
+/// but MIR optimizations subsequently remove all of its [`CoverageKind::CounterIncrement`]
+/// statements (e.g. because bb0 is unreachable), then we won't generate any
+/// `llvm.instrprof.increment` intrinsics. LLVM will think the function is not
+/// instrumented, and it will disappear from coverage mappings and coverage reports.
+///
+/// This pass detects when that has happened, and re-inserts a dummy counter-increment
+/// statement so that LLVM knows to treat the function as instrumented.
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/coverage/repair.rs:15]: I wonder if this should be done by codegen directly. When starting codegen for a mono-item, check if `function_coverage_info` is set, and add this logic?

Comment (Zalathar) [reply compiler/rustc_mir_transform/src/coverage/repair.rs:15]: Yeah, that sounds reasonable. I'll have to track down the right place to add a hook, but once I do the result should be more self-contained in codegen.

**Comment by Zalathar** [comment]
Hmm, I was investigating this some more, and found that I couldn't get my tests to fail in the way they were failing before.

That means I can't be confident that my fix is working, or even necessary.

I'll put this on hold until I figure out what's going on.

@rustbot author

**Comment by Dylan-DPC** [comment]
@Zalathar any updates on this? Thanks

**Comment by Zalathar** [comment]
@Dylan-DPC This has been a low priority for me, so no updates unfortunately.

**Comment by Zalathar** [comment]
I found a much nicer way to do this:

- When looking for unused functions, we already need to scan through the body of each codegenned function, looking for coverage statements that were inlined from other functions.
- During this scan, we can also keep track of whether we’ve seen any of the function’s own coverage statements.
- If a function was instrumented but has no coverage statements, we can just treat it as an unused function.

**Comment by Zalathar** [comment]
Closing in favour of #122860.

**Pull request closed without merging**